### PR TITLE
p2p: make block download logic aware of limited peers threshold

### PR DESCRIPTION
**Pull request by furszy**
Even when the node believes it has IBD completed, need to avoid
requesting historical blocks from network-limited peers.
Otherwise, the limited peer will disconnect right away.

The simplest scenario could be a node that gets synced, drops
connections, and stays inactive for a while. Then, once it re-connects
(IBD stays completed), the node tries to fetch all the missing blocks
from any peer, getting disconnected by the limited ones.

Note:
Can verify the behavior by cherry-picking the test commit alone on
master. It will fail there.

### CODE REVIEW

**Review by mzumsande — COMMENTED** [review]
Maybe getting disconnected could be beneficial in this rare situation? If our tip is that much behind but we're not in IBD, we can't help them and they can't help us, so our priority should be to get the missing blocks asap, and getting disconnected by useless peers seems like it's not such a bad thing towards that goal.

What would be bad is if all our current peers were limited, and we wouldn't try to request the needed blocks from anyone, but also wouldn't try to exchange them for better peers, so that we would be stuck.

**Review by furszy — COMMENTED** [review]
> Maybe getting disconnected could be beneficial in this rare situation? If our tip is that much behind but we're not in IBD, we can't help them and they can't help us, so our priority should be to get the missing blocks asap, and getting disconnected by useless peers seems like it's not such a bad thing towards that goal.

Agree but, while the outcome might be the desired one, the approach to achieve it isn't the best. It seems unwise to have the node sending a message through the wire, knowing it will fail, just to force the other peer to disconnect.
The node should be smarter than that and not rely on external factors to take decisions.

Moreover, in cases where the other peer chooses to ignore the message without disconnecting (maybe due to using different software than bitcoin-core), the node will end up waiting for a response that will never arrive, wasting time and resources, which could have been easily avoided.

And also, while the node has more connection slots available, I don't see why should disconnect a peer that is behaving properly.

Instead, I think that we should try to keep processes separated. The eviction process should be in charge of disconnecting peers that aren't/will not provide any meaningful data. And the block downloading logic should only be responsible of deciding which blocks request to each peer and which ones not.

**Comment by naumenkogs** [comment]
> What would be bad is if all our current peers were limited, and we wouldn't try to request the needed blocks from anyone, but also wouldn't try to exchange them for better peers, so that we would be stuck.

I agree with this part. I think the solution should be expanded to give a chance in such cases, and it should be merged within the same PR.

**Comment by luke-jr** [comment]
Agree with @mzumsande and @naumenkogs. Maybe not the same PR if it's substantially different, but the new eviction logic should be merged before this fix.

**Review by furszy — COMMENTED** [review]
> What would be bad is if all our current peers were limited, and we wouldn't try to request the needed blocks from anyone, but also wouldn't try to exchange them for better peers, so that we would be stuck.

Right now, this will hardly happen. The reason is `CheckForStaleTipAndEvictPeers`. The logic is as follows:

When the node doesn't update the tip for a period longer than 30 minutes (`TipMayBeStale()`), it triggers the extra outbound connection process. Which instructs the net layer to bypass the maximum outbound connection restriction and create another full relay outbound connection to sync up the missing blocks.

Then, considering the worst-case scenario: when the extra outbound connection is also a limited peer. In that case, 45 seconds after connecting to the extra peer (scheduler task timeout), `CheckForStaleTipAndEvictPeers` will call `EvictExtraOutboundPeers`, which will evict the peer that least recently announced a block header.
Then, in case of no chain movement; the process starts again, adding an extra outbound connection, and so on, until it finds a peer that can provide the missing blocks.

Note:
While this sooner or later corrects the stalling situation, it's not ideal to continue establishing connections to limited peers, on the extra outbound connection process, when the peer is stalled. So, to address this issue, have created #28170. Which basically disallows connections to limited peers when the node is behind enough to know that limited peers will not provide any of the required historical blocks (although they will still provide valid headers).

**Review by furszy — COMMENTED** [review]
rebased. Conflicts solved.

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1443,30 +1444,46 @@ void PeerManagerImpl::FindNextBlocks(std::vector<const CBlockIndex*>& vBlocks, c
...
+                // We reached the end of the window.
+                if (vBlocks.size() == 0 && waitingfor != peer.m_id) {
+                    // We aren't able to fetch anything, but we would be if the download window was one larger.
+                    if (nodeStaller) *nodeStaller = waitingfor;
+                }
+                return;
+            }
+
+            // Don't request blocks that go further than what limited peers can provide
+            if (is_limited_peer && (state->pindexBestKnownBlock->nHeight - pindex->nHeight > static_cast<int>(NODE_NETWORK_LIMITED_MIN_BLOCKS) + 2 /* add two blocks buffer extension for possible races */)) {
```
Comment (andrewtoth) [inline src/net_processing.cpp:1480]: I don't think we want to keep the + 2 buffer here. This means we would potentially still request blocks 2 blocks after 288, so up to 290. The buffer for sending requested blocks up to 288 + 2 is there for when the sender sees new blocks but the requester has not yet seen it. In this case we are requesting so we should respect the 288 limit.

File: test/functional/p2p_node_network_limited.py [hunk 2]
```diff
@@ -46,6 +49,44 @@ def setup_network(self):
...
+        for node in self.nodes:
+            assert not node.getblockchaininfo()['initialblockdownload']
+
+        # Disconnect full_node from the other two peers (the node will remain out of IBD)
+        self.disconnect_nodes(0, 2)
+        self.disconnect_nodes(1, 2)
+        assert_equal(len(full_node.getpeerinfo()), 0)
+
+        # Mine blocks and sync the pruned node. Surpass the NETWORK_NODE_LIMITED threshold.
+        self.generate(miner, MIN_BLOCKS_TO_KEEP + 12, sync_fun=self.no_op)
```
Comment (vasild) [inline test/functional/p2p_node_network_limited.py:79]: Is the intention here to mimic `NODE_NETWORK_LIMITED_MIN_BLOCKS` from the C++ source code, which is `288`? `MIN_BLOCKS_TO_KEEP` is a different constant defined in both the C++ and Python code which happens to be also `288`. Should `NODE_NETWORK_LIMITED_MIN_BLOCKS` be defined in the Python code (equal to 288, but not necessary equal or tied to `MIN_BLOCKS_TO_KEEP`)?

File: test/functional/p2p_node_network_limited.py [hunk 3]
```diff
@@ -46,6 +49,44 @@ def setup_network(self):
...
+        self.sync_blocks([miner, pruned_node])
```
Comment (vasild) [inline test/functional/p2p_node_network_limited.py:80]: Here `sync_blocks()` is used and above `self.wait_until(lambda: node.getbestblockhash() == ...` is used to achieve the same purpose. For consistency maybe use only one of the methods.

File: test/functional/p2p_node_network_limited.py [hunk 4]
```diff
@@ -46,6 +49,44 @@ def setup_network(self):
...
+
+        # Connect full_node to prune_node and check peers don't disconnect right away.
+        # (they will disconnect if full_node, which is chain-wise behind, request blocks
+        #  older than MIN_BLOCKS_TO_KEEP)
+        self.connect_nodes(0, 2)
+        assert_equal(len(full_node.getpeerinfo()), 1)
+        time.sleep(3)  # Wait for a few seconds to ensure peers remain connected
+        assert_equal(len(full_node.getpeerinfo()), 1)
```
Comment (vasild) [inline test/functional/p2p_node_network_limited.py:88]: This sleep is non-deterministic. The good thing is that it will not cause sporadic test failures, but still, it could occasionally report success even if the full node requests an old block and gets disconnected (false positive). Also it adds unconditional 3 seconds to the tests execution time.

Is it possible to lookup something in the log or via RPC to cross check that the full node has made a decision not to request blocks from the pruned node?

**Review by vasild — COMMENTED** [review]
Approach ACK c1612ea1f0

File: src/net_processing.cpp [hunk 5]
```diff
@@ -1443,30 +1444,46 @@ void PeerManagerImpl::FindNextBlocks(std::vector<const CBlockIndex*>& vBlocks, c
                 // We consider the chain that this peer is on invalid.
                 return;
             }
+
             if (!CanServeWitnesses(peer) && DeploymentActiveAt(*pindex, m_chainman, Consensus::DEPLOYMENT_SEGWIT)) {
                 // We wouldn't download this block or its descendants from this peer.
                 return;
             }
+
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/net_processing.cpp:1452]: Thx, done as suggested.

File: test/functional/p2p_node_network_limited.py (hunk 3 above)
Comment (furszy) [reply test/functional/p2p_node_network_limited.py:78]: Thx, done as suggested.

**Review by furszy — COMMENTED** [review]
Updated per feedback, thanks for the review!

Fixed the two blocks buffer extension (now is reduction). Small [diff](https://github.com/bitcoin/bitcoin/compare/c1612ea1f050b3daede6a181464bbbce6e254b2d..9eba981808a56eee61a20b8378447bd0972ad589).

File: src/net_processing.cpp [hunk 6]
```diff
@@ -1443,30 +1444,46 @@ void PeerManagerImpl::FindNextBlocks(std::vector<const CBlockIndex*>& vBlocks, c
...
+                if (vBlocks.size() == 0 && waitingfor != peer.m_id) {
+                    // We aren't able to fetch anything, but we would be if the download window was one larger.
+                    if (nodeStaller) *nodeStaller = waitingfor;
+                }
+                return;
+            }
+
+            // Don't request blocks that go further than what limited peers can provide
+            if (is_limited_peer && (state->pindexBestKnownBlock->nHeight - pindex->nHeight > static_cast<int>(NODE_NETWORK_LIMITED_MIN_BLOCKS) + 2 /* add two blocks buffer extension for possible races */)) {
+                return;
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/net_processing.cpp:1481]: Ok, lets use an example (from the newly added test):

`vToFetch = [2 ... 129]`
`state->pindexBestKnownBlock->nHeight = 301`
So the peer's height is `301`. We can assume they can provide blocks `[13 ... 301]`.
At the first iteration of the loop `pindex->nHeight` is 2 (`vToFetch[0]`).
The code ends up with `return;` because `301 - 2 > 286` and we never request any blocks from that peer.

Shouldn't we request blocks `[13 ... 301]` instead?

File: test/functional/p2p_node_network_limited.py [hunk 7]
```diff
@@ -46,6 +49,42 @@ def setup_network(self):
...
+    def test_avoid_requesting_old_blocks(self):
+        self.log.info("Test node not requesting blocks beyond limited peer threshold")
+        # The request of blocks further than the NETWORK_NODE_LIMITED threshold causes a direct disconnection.
+        pruned_node = self.nodes[0]
+        miner = self.nodes[1]
+        full_node = self.nodes[2]
+
+        # Connect and generate block to ensure IBD=false
+        self.connect_nodes(1, 2)
+        self.connect_nodes(0, 2)
```
Comment (vasild) [inline test/functional/p2p_node_network_limited.py:61]: The test would fail if run standalone without the preceding tests. The failure is on line 78: `self.sync_blocks([miner, pruned_node])`. That is because the test relies on the pruned node being connected to the miner which is indeed the case - some leftover from previous tests. This diff makes it succeed when run standalone:

```diff
--- i/test/functional/p2p_node_network_limited.py
+++ w/test/functional/p2p_node_network_limited.py
@@ -54,25 +54,24 @@ class NodeNetworkLimitedTest(BitcoinTestFramework):
         # The request of blocks further than the NETWORK_NODE_LIMITED threshold causes a direct disconnection.
         pruned_node = self.nodes[0]
         miner = self.nodes[1]
         full_node = self.nodes[2]
 
         # Connect and generate block to ensure IBD=false
+        self.connect_nodes(0, 1)
         self.connect_nodes(1, 2)
-        self.connect_nodes(0, 2)
         self.generate(miner, 1, sync_fun=self.no_op)
 
         # Verify all peers are at the same height
         self.sync_blocks([miner, pruned_node])
                                      
         # Verify peers are out of IBD
         for node in self.nodes:
             assert not node.getblockchaininfo()['initialblockdownload']    
 
         # Disconnect full_node from the other two peers (the node will remain out of IBD)
-        self.disconnect_nodes(0, 2)
         self.disconnect_nodes(1, 2)                                       
         assert_equal(len(full_node.getpeerinfo()), 0)                     
 
         # Mine blocks and sync the pruned node. Surpass the NETWORK_NODE_LIMITED threshold.
```

It is also shorter and I think more clear.

File: src/net_processing.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (furszy) [reply src/net_processing.cpp:1481]: Hmm, that will only happen when the node is close to the tip, because of the 1024 block download window. More precisely, when the node is within 1312 blocks from the tip (window + limited peers threshold). Nice observation!.
Will modify it and add coverage for it.

**Review by furszy — COMMENTED** [review]
Updated per feedback, thanks for the review @vasild!

The node will now request blocks within the limited peers threshold when the block download window contemplates them and not only when the previous blocks arrived or are in-flight. Added test coverage exercising this behavior and verifying that the node can get synced even when the blocks arrived out of order.

File: test/functional/p2p_node_network_limited.py (hunk 7 above)
Comment (furszy) [reply test/functional/p2p_node_network_limited.py]: Ended up doing it in a slightly different way. Using `setnetworkactive` to isolate the full node from the other two. I think that is clearer in this way.

File: test/functional/p2p_node_network_limited.py (hunk 4 above)
Comment (furszy) [reply test/functional/p2p_node_network_limited.py]: Removed the unconditional wait on the last update :).

File: test/functional/p2p_node_network_limited.py (hunk 2 above)
Comment (furszy) [reply test/functional/p2p_node_network_limited.py]: Added `NODE_NETWORK_LIMITED_MIN_BLOCKS` on the last update.

File: src/net_processing.cpp [hunk 8]
```diff
@@ -1443,30 +1444,46 @@ void PeerManagerImpl::FindNextBlocks(std::vector<const CBlockIndex*>& vBlocks, c
...
+                if (vBlocks.size() == 0 && waitingfor != peer.m_id) {
+                    // We aren't able to fetch anything, but we would be if the download window was one larger.
+                    if (nodeStaller) *nodeStaller = waitingfor;
+                }
+                return;
+            }
+
+            // Don't request blocks that go further than what limited peers can provide
+            if (is_limited_peer && (state->pindexBestKnownBlock->nHeight - pindex->nHeight > static_cast<int>(NODE_NETWORK_LIMITED_MIN_BLOCKS) - 2 /* two blocks buffer for possible races */)) {
+                continue;
```
Comment (vasild) [inline src/net_processing.cpp:1481]: I think there is an off-by-one error here, should this be `>=` instead of `>`? Lets say the limited peer is at height `H` then the intention of this is to request blocks `(H-286, H]` from it, a total of `286` blocks. However it would not `continue` for height `H-286` and would request that block too.

File: test/functional/p2p_node_network_limited.py [hunk 9]
```diff
@@ -46,6 +50,65 @@ def setup_network(self):
...
+        # expected to be downloaded from the limited peer.
+        tip_height = pruned_node.getblockcount()
+        self.wait_until(lambda: not try_rpc(-1, "Block not found", full_node.getblock, pruned_node.getbestblockhash()))
+        for height in range(tip_height, start_height_full_node + limited_threshold_buffer, -1):
+            full_node.getblock(pruned_node.getblockhash(height))  # just need to assert it does not raise an exception
+
+        # Verify all blocks that should not have been requested.
+        for move in range(1, limited_threshold_buffer):
+            height = start_height_full_node + move
+            assert_raises_rpc_error(-1, "Block not found on disk", full_node.getblock, pruned_node.getblockhash(height))
```
Comment (vasild) [inline test/functional/p2p_node_network_limited.py:104]: The full node is at height `1` and the pruned node is at height `289`. The first loop verifies that the full node has blocks `[4, 289]` (`286` blocks in total), which is ok. But then it only checks that the full node does not have block `2`. What about block `3`? It is not checked and should not have been downloaded but it has been downloaded.

To avoid the gap the range in the second loop should be `range(1, limited_threshold_buffer + 1)` or even:

```python
for height in range(start_height_full_node + 1, start_height_full_node + 1 + limited_threshold_buffer):
```

or have one loop instead of two:

```python
for height in range(start_height_full_node + 1, tip_height + 1):
    if height <= tip_height - (NODE_NETWORK_LIMITED_MIN_BLOCKS - 2):
        assert_raises_rpc_error(-1, "Block not found on disk", full_node.getblock, pruned_node.getblockhash(height))
    else:
        full_node.getblock(pruned_node.getblockhash(height))
```

Comment (furszy) [reply test/functional/p2p_node_network_limited.py:104]: Done as suggested. Thanks.

File: src/net_processing.cpp (hunk 8 above)
Comment (furszy) [reply src/net_processing.cpp:1481]: Sounds good. Corrected.

**Review by vasild — APPROVED** [review]
ACK a0687ffd568b0984a8b1e51fd0061f87f25e95d7

Thanks!

**Review by andrewtoth — APPROVED** [review]
ACK a0687ffd568b0984a8b1e51fd0061f87f25e95d7

File: test/functional/p2p_node_network_limited.py [hunk 10]
```diff
@@ -46,6 +50,63 @@ def setup_network(self):
...
+        for node in self.nodes:
+            assert not node.getblockchaininfo()['initialblockdownload']
+
+        # Isolate full_node (the node will remain out of IBD)
+        full_node.setnetworkactive(False)
+        self.wait_until(lambda: len(full_node.getpeerinfo()) == 0)
+
+        # Mine blocks and sync the pruned node. Surpass the NETWORK_NODE_LIMITED threshold.
+        # Blocks deeper than the threshold are considered "historical blocks"
+        num_historial_blocks = 12
```
[the remark this answers is not in the document]
Comment (andrewtoth) [reply test/functional/p2p_node_network_limited.py:74]: Ah yes, it makes sense what `num_historical_blocks` purpose is. The nit is why the number 12 was chosen specifically. I suppose to just make it an even 300, and it can be arbitrary?

**Comment by naumenkogs** [comment]
The code looks alright, but i'd still rather have [this](https://github.com/bitcoin/bitcoin/pull/28120#pullrequestreview-1557211930) addressed first. It looks like even #28170 doesn't directly handle this, but rather prepare the codebase for another PR handling it?

**Review by furszy — COMMENTED** [review]
Sorry for the delay @naumenkogs. 've been on #28170 before returning here. All good now, it is ready to go.

> The code looks alright, but i'd still rather have [this](https://github.com/bitcoin/bitcoin/pull/28120#pullrequestreview-1557211930) addressed first. It looks like even #28170 doesn't directly handle this, but rather prepare the codebase for another PR handling it?

The concern was:
All connections are from limited peers, we are lagging behind and we no longer disconnect from them due the lack of the historical block request disconnection misfeature.

With #28170, when this happens, the node will detect that it is lagging behind and update the desirable connection service flags accordingly, ensuring that the extra outbound connection process only connects to full nodes (unlike now, where the node can still connect to another limited peer, etc). Then, once the connection is established, the node will recover from the stalling situation.

Essentially, instead of resolving the situation by evicting a good limited peer connection (which could be self-defeating in some cases), we will solve it by establishing a new connection. With this, I'm not saying that we couldn't actually do both (it wouldn't be bad to reserve some outbound slots for full nodes and/or evict the least helpful limited peer when the node is stalled), but I think the concern is no longer a blocker after #28170.

File: src/net_processing.cpp [hunk 11]
```diff
@@ -1475,6 +1476,11 @@ void PeerManagerImpl::FindNextBlocks(std::vector<const CBlockIndex*>& vBlocks, c
                 return;
             }
 
+            // Don't request blocks that go further than what limited peers can provide
+            if (is_limited_peer && (state->pindexBestKnownBlock->nHeight - pindex->nHeight >= static_cast<int>(NODE_NETWORK_LIMITED_MIN_BLOCKS) - 2 /* two blocks buffer for possible races */)) {
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/net_processing.cpp:1480]: I'm not sure what you are referring to. Could you rephrase it please.
You want to update `pindexLastCommonBlock` to `pindex`?

**Review by vasild — APPROVED** [review]
ACK c5b5843d8f10d96f76ee6b95f2b1b1b4ce755f75

File: src/net_processing.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sicelo7) [reply src/net_processing.cpp:1482]: Ye

File: test/functional/p2p_node_network_limited.py [hunk 12]
```diff
@@ -54,6 +58,63 @@ def setup_network(self):
...
+    def test_avoid_requesting_historical_blocks(self):
+        self.log.info("Test full node does not request blocks beyond the limited peer threshold")
+        pruned_node = self.nodes[0]
+        miner = self.nodes[1]
+        full_node = self.nodes[2]
+
+        # Connect and generate block to ensure IBD=false
+        self.connect_nodes(1, 0)
+        self.connect_nodes(1, 2)
+        self.generate(miner, 1)
```
[the remark this answers is not in the document]
Comment (furszy) [reply test/functional/p2p_node_network_limited.py:70]: > I don't understand this part of the test. Why do we connect nodes 1 and 2, just to disconnect again without doing anything in between? Why is the generate call necessary?

To ensure all nodes are out of IBD. The idea was to isolate this test case from the previous cases executed along the file. It was done in this way to be able to run this test case at any point of the file. This was vasild request because he was running it first, commenting the rest of the cases to speed the process up.

The connection requirement is because the test overrides `setup_network`, which skips the framework's connections creation process (peers start the test with no connections).

> shouldn't we then also at least add a sync_blocks to ensure it propagates to all nodes before disconnecting?

The `self.generate(miner, 1)` function calls `sync_blocks()` internally.

[the remark this answers is not in the document]
Comment (mzumsande) [reply test/functional/p2p_node_network_limited.py:70]: ok, makes sense!

File: src/net_processing.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (andrewtoth) [reply src/net_processing.cpp:1510]: I think what @naumenkogs is referring to here is that if this condition is true then `pindex` will hit this condition on the first `pindex` in the loop and every `pindex` up until this condition is no longer true. This can be more efficient by just bumping `pindex` to the first `pindex` that is past the network limited blocks threshold.

So, we can either change the loop from range based to regular for-loop and advance the index, or move this check out of this loop and into the loop on line 1464 and break early if the condition is met.

**Review by mzumsande — COMMENTED** [review]
Code Review ACK c5b5843d8f10d96f76ee6b95f2b1b1b4ce755f75

File: src/net_processing.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (furszy) [reply src/net_processing.cpp:1510]: > I think what @naumenkogs is referring to here is that if this condition is true then `pindex` will hit this condition on the first `pindex` in the loop and every `pindex` up until this condition is no longer true. This can be more efficient by just bumping `pindex` to the first `pindex` that is past the network limited blocks threshold.
>
> So, we can either change the loop from range based to regular for-loop and advance the index, or move this check out of this loop and into the loop on line 1464 and break early if the condition is met.

Ok, we are on the same page now. Thanks. Let's leave it for a follow-up. It would be nice to re-think this function and benchmark it properly. In a first glance, I think we would be optimizing the function for the less likely case and may degrade performance for the likely ones.

**Review by pinheadmz — APPROVED** [review]
ACK c5b5843d8f10d96f76ee6b95f2b1b1b4ce755f75

Built and ran all functional and unit tests on x86 Ubuntu. Confirmed test fails on master, passes with the patch in the branch. Reviewed code and existing discussion. Regarding the decision to keep the pruning peer connected at all, I agree with the author that the affected code is only responsible for requesting available blocks from a peer and it is reasonable to take the limited flag into account. I like the clean up in the first commit as well, it's easier to read the loop with `continue`.

For sanity I added checks in the test to confirm blocks from the pruned node were "found" by the full node but had `'confirmations': -1,` which makes sense because there is a gap between the full node's chain tip and the blocks available from the pruning peer.


<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK c5b5843d8f10d96f76ee6b95f2b1b1b4ce755f75
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmXs7UsACgkQ5+KYS2KJ
yTpnUg//b9HSUh/s1VJRWSoX23pcmFJtmeA4ndd5dGM+Q2RbyBZenI4ZwVlkunxy
l5B+U/gpD7cL/6iZsV6Eph/ouX9Om8NKQcxMn0TXzvOBYcByCZB++r5oNwfeKohu
aUJWRce9MYDj4GtdbW7p5DwcsKaX6nC8oOnxfv9ivP473F7BDjnYZpX7F0nKsrx7
b5vKMWCY7fXxmHpe0y0FnWVAJ10hvIQFD6I0EMO1mQpq14U3NUcG+Axfm6kCrJRE
EbeAKtxz/tITNNiN5AZg2NdgnmOR1MaOatkOuCxWgMer+N2EOWZgeW+/42Iy0Isz
hOnv9BQi4MwewY/I5ruM7Yx2oghUEpIkIhBX8/JPeA9iD2pnVIcXi6Xkaawd31C1
Xl7OMpX1mOVtPjNPvmYxpbFjze6BhGpaa3h7UJFSBzqhFm0gpe70f+hza4HbXUQj
MihsnKASJPmCDLFY/kGddad/E+yJ5QNxSHhLqEugkWPGPnIa4CHyd91uU5qQwwOv
H1sN9hAlNYguDZs79jPY32yIXgpufQyUWQA1+emQpDQun68jHjTxTdgUFEbvlIDU
Nl+fxsY9Fqj4D3HpdnnwIUcX1W68itMhPPBQUDu3R1abdUHMc/ntE4gwr6BP3LcL
AFv+d4aflQrEG9RntgmkyUSnYw6VIj9INlQMS6ql4r70Kh/6Bhk=
=C8tL
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Comment by achow101** [comment]
ACK c5b5843d8f10d96f76ee6b95f2b1b1b4ce755f75

File: src/net_processing.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (jonatack) [reply src/net_processing.cpp:1510]: > Let's leave it for a follow-up. It would be nice to re-think this function

Any update?
